benches/aead.rs

```diff
@@ -0,0 +1,162 @@
+impl NonceSequence {
+    const fn new() -> Self {
+        Self(0)
+    }
+}
+
+impl aead::NonceSequence for NonceSequence {
+    fn advance(&mut self) -> Result<aead::Nonce, error::Unspecified> {
+        let mut result = [0u8; aead::NONCE_LEN];
+        result[4..].copy_from_slice(&u64::to_be_bytes(self.0));
+        self.0 = self.0.checked_add(1).ok_or(error::Unspecified)?;
+        Ok(aead::Nonce::assume_unique_for_key(result))
+    }
+}
+
+macro_rules! bench {
+    ( $benchmark_name:ident, $algorithm:expr, $chunk_len:expr, $ad:expr ) => {
+        pub(super) fn $benchmark_name(c: &mut criterion::Criterion) {
+            use super::{NonceSequence, NONCE};
+            use criterion::{black_box, BatchSize};
+            use ring::{
+                aead::{self, BoundKey},
+                rand::{SecureRandom, SystemRandom},
+            };
+
+            let rng = SystemRandom::new();
+
+            let mut key_bytes = vec![0u8; $algorithm.key_len()];
+            rng.fill(&mut key_bytes).unwrap();
+
+            const FUNC_NAME_BASE: &str =
+                concat!(stringify!($benchmark_name), "_", stringify!($algorithm));
+            let func_name_cleaned = FUNC_NAME_BASE.replace("&aead::", "");
+
+            {
+                let key = aead::UnboundKey::new($algorithm, &key_bytes).unwrap();
+                let mut key = aead::SealingKey::new(key, NonceSequence::new());
+
+                let mut in_out = black_box(vec![0u8; $chunk_len + $algorithm.tag_len()]);
+                let enc_func_name = format!("encrypt_{}", func_name_cleaned);
+                c.bench_function(&enc_func_name, |b| {
+                    b.iter(|| {
+                        let aad = black_box(aead::Aad::from($ad));
+                        key.seal_in_place_append_tag(aad, &mut in_out).unwrap();
+                    })
+                });
+            }
+
+            {
+                let key =
+                    aead::LessSafeKey::new(aead::UnboundKey::new($algorithm, &key_bytes).unwrap());
+
+                let nonce = aead::Nonce::assume_unique_for_key(NONCE);
+
+                let ciphertext = {
+                    let mut in_out = vec![0u8; $chunk_len + $algorithm.tag_len()];
+                    let aad = aead::Aad::from($ad);
+                    key.seal_in_place_append_tag(nonce, aad, &mut in_out)
+                        .unwrap();
+                    in_out
+                };
+
+                let dec_func_name = format!("decrypt_{}", func_name_cleaned);
+                c.bench_function(&dec_func_name, |b| {
+                    b.iter_batched(
+                        || ciphertext.clone(),
+                        |mut in_out| {
+                            // Optimizes out
+                            let nonce = aead::Nonce::assume_unique_for_key(NONCE);
+
+                            let aad = black_box(aead::Aad::from($ad));
+                            key.open_in_place(nonce, aad, &mut in_out).unwrap();
```

Comment:
Most/all the comments for the sealing benchmark apply to this opening benchmark too.

Reply:
These should also be addressed since I modified both as part of reviewing the previous comments.